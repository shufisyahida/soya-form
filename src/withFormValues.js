import { set } from "lodash";
import React from "react";
import { connect } from "react-redux";
import { STATE_NAME } from "./_constants";
import createSoyaFormSelector from "./_selectors";
import { getDisplayName } from "./_utils";

export default (FORM_ID, fieldNames) => Component => {
  const mapStateToProps = state => {
    let formValues = {};
    if (state[STATE_NAME]) {
      const soyaFormSelector = createSoyaFormSelector(state, FORM_ID);
      formValues = fieldNames.reduce((formValues, fieldName) => {
        const fieldPath = Array.isArray(fieldName) ? fieldName : [fieldName];
        const value = soyaFormSelector.getFieldValue(fieldPath);
        set(formValues, fieldPath, value);
        return formValues;
      }, {});
    }
    return { formValues };
  };

  return connect(mapStateToProps)(
    class extends React.Component {
      static displayName = getDisplayName("WithFormValues", Component);

      render() {
        const { formValues, ...props } = this.props;
        delete props.dispatch;
        return <Component formValues={formValues} {...props} />;
      }
    }
  );
};
