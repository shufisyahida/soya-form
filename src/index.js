import createForm from './createForm';
import createField from './createField';
import createRepeatable from './createRepeatable';
import createSelector from './_selectors';
import withForm from './withForm';
import withFormValues from './withFormValues';

export {
  createForm,
  createField,
  createRepeatable,
  createSelector,
  withForm,
  withFormValues,
};
