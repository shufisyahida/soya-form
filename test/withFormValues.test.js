import React from "react";
import { Provider } from "react-redux";
import TestRenderer from "react-test-renderer";
import { createConfigureStore } from "soya-next/redux";
import { createField, createForm, createRepeatable, withFormValues } from "../src";

let store;
const configureStore = createConfigureStore();

describe("withFormValues", () => {
  beforeEach(() => {
    store = configureStore();
  });

  const ArrayField = createRepeatable(({ name }) => (
    <div>
      <InputField name={name("title")} />
      <InputField name={name("start")} />
      <InputField name={name("end")} />
    </div>
  ));
  const InputField = createField(({ handleChange, name, value }) => (
    <input
      type="text"
      name={name}
      value={value}
      onChange={e => handleChange(e.target.value)}
    />
  ));
  const FORM_ID = "mockForm";
  const Component = ({ formValues }) => JSON.stringify(formValues);
  const FlatFormValues = withFormValues(FORM_ID, ["firstName", "lastName"])(
    Component
  );
  const DeepFormValues = withFormValues(FORM_ID, ["firstName", ["address", "streetName"], ["experiences", 0, "title"]])(
    Component
  );
  const MockForm = createForm(FORM_ID)(() => (
    <form>
      <InputField name="firstName" />
      <InputField name="middleName" />
      <InputField name="lastName" />
      <InputField name={["address", "streetName"]} />
      <InputField name={["address", "streetNo"]} />
      <InputField name={["address", "city"]} />
      <InputField name={["address", "countryCode"]} />
      <InputField name={["address", "contact"]} />
      <ArrayField name="experiences" />
      <FlatFormValues />
      <DeepFormValues />
    </form>
  ));

  it("should have formValues prop", () => {
    const renderer = TestRenderer.create(
      <Provider store={store}>
        <MockForm />
      </Provider>
    );
    const { root } = renderer;
    expect(root.findAllByType(Component).map(component => component.props)).toMatchSnapshot();
  });
});
