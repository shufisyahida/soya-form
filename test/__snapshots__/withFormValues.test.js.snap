// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withFormValues should have formValues prop 1`] = `
Array [
  Object {
    "formValues": Object {
      "firstName": null,
      "lastName": null,
    },
  },
  Object {
    "formValues": Object {
      "address": Object {
        "streetName": null,
      },
      "experiences": Array [
        Object {
          "title": null,
        },
      ],
      "firstName": null,
    },
  },
]
`;
